### PR TITLE
Enhance profiler results

### LINKED ISSUE #898: Add toString to KeySliceQuery for profile
Actually, when you use profile you do not know if your query match a vertex-centric indexes or not because.

> GraphStep(vertex,[3539008])                                            1           1           0.107     0.37
> JanusGraphVertexStep([rating.gt(3.0)])                                 6           6          29.252    99.63
>     \_condition=(rating > 3.0 AND type[battled])
>     \_isFitted=false
>     \_vertices=1
>     **\_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81b60e15**
>     \_orders=[DESC(time)]
>     \_isOrdered=false
>   optimization                                                                                 0.274
>                                             >TOTAL                     -           -          29.359

### PR DESCRIPTION
**Pull request by li-boxuan**
1. Add toString() method to SliceQuery to indicate which vertex-centric
index is used in edge queries.
2. Indicate "GraphCentricQuery" and "VertexCentricQuery" in profiler if applicable.
Previously they were incorrectly merged, causing corrupted annotations.
3. Indicate "constructGraphCentricQuery" phase in profiler rather than a vague
"optimization" annotation, if applicable.
4. Fix corrupted annotations caused by nested profile flattening issue. Now
AND/OR nesting is flattened only if it does not have siblings.
5. Remove backend-query subgroup to reduce nesting levels in profiler result.
Now whether a backend-query is fired can be inferred from "isCached" annotation.
6. Remove fullscan subgroup to reduce nesting levels in profiler result. Now
whether a full scan is done can be inferred from "fullscan" annotation.
Fixes #898, #2283, #2285

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java [hunk 1]
```diff
@@ -5661,6 +5662,161 @@ public void testNotHas() {
...
+        JanusGraphVertex v = tx.addVertex();
+        JanusGraphVertex o = tx.addVertex();
+        v.addEdge("friend", o, "time", 100);
+        v.addEdge("friend-no-index", o, "time", 100);
+        tx.commit();
+
+        // satisfied by a single vertex-centric query which is satisfied by union of two edge queries
+        newTx();
+        Metrics mMultiLabels = tx.traversal().V(v).outE("friend", "friend-no-index").has("time", 100)
+            .profile().next().getMetrics(1);
```
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5790]: With this PR:
```
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[v[4128]])                                            1           1           1.086    21.07
JanusGraphVertexStep([time.eq(100)])                                   2           2           4.068    78.93
  VertexCentricQuery                                                                           0.000
    \_condition=(time = 100 AND (type[friend] OR type[friend-no-index]))
    \_orders=[]
    \_vertices=1
    optimization                                                                               2.099
    OR-query                                                                                   0.000
    \_isFitted=true
    \_isOrdered=true
    \_query=2069:byTime:SliceQuery[0xB0E0FF7FFFFF9B,0xB0E0FF7FFFFF9C)@2147483647
      backend-query                                                    1                       0.045
    \_query=SliceQuery[0xB0E0FF7FFFFF9B,0xB0E0FF7FFFFF9C)@2147483647
    OR-query                                                                                   0.000
    \_isFitted=false
    \_isOrdered=true
    \_query=friend-no-index:SliceQuery[0x7180,0x7181)@2147483647
      backend-query                                                    1                       0.040
    \_query=SliceQuery[0x7180,0x7181)@2147483647
                                            >TOTAL                     -           -           5.154        -
```

Without this PR (0.6.0-SNAPSHOT):
```
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[v[4136]])                                            1           1           0.984    17.13
JanusGraphVertexStep([time.eq(100)])                                   2           2           4.762    82.87
    \_condition=(time = 100 AND (type[friend] OR type[friend-no-index]))
    \_orders=[]
    \_isFitted=false
    \_isOrdered=true
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@812ad47d
    \_vertices=1
  optimization                                                                                 3.297
  backend-query                                                        1                       0.054
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@d5c81d3d
  backend-query                                                        1                       0.026
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@812ad47d
                                            >TOTAL                     -           -           5.746        -
```

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java [hunk 2]
```diff
@@ -79,6 +79,6 @@ public boolean subsumes(KeySliceQuery oth) {
 
     @Override
     public String toString() {
-        return String.format("KeySliceQuery(key: %s, start: %s, end: %s, limit:%d)", key, getSliceStart(), getSliceEnd(), getLimit());
+        return "KeySliceQuery(" + key + ")[" + getSliceStart() + "," + getSliceEnd() + ")@" + getLimit();
```
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java:82]: Why is the `[` closed by `)`?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java [hunk 3]
```diff
@@ -242,6 +239,9 @@ public GraphCentricQuery constructQueryWithoutProfile(final ElementCategory resu
         Preconditions.checkNotNull(resultType);
         if (limit == 0) return GraphCentricQuery.emptyQuery(resultType);
 
+        if (this.globalConstraints.isEmpty()) {
```
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java:242]: What is the reason for this change?

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java (hunk 2 above)
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java:82]: This is to indicate this is a half-closed interval: [sliceStart, sliceEnd) means sliceStart is inclusive while sliceEnd is exclusive. The old way is clear but a bit lengthy.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java (hunk 3 above)
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java:242]: This is because now `JanusGraphStep` calls `constructQueryWithoutProfile` directly, while previously, it calls `constructQuery` which in turn calls `constructQueryWithoutProfile`. Thus I have to move this logic. Anyway, I think the right place for this logic is indeed the `constructQueryWithoutProfile` method.

The reason for calling `constructQueryWithoutProfile` directly is because we don't want the graph centric query to be profiled. `JanusGraphStep` always 1) tries to construct a single graph centric query and see if it can fulfill the constraints. If not, it 2) constructs a few graph centric queries and combines results together. If Step 1) does not succeed, we don't want it to appear in the profiler result.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java [hunk 4]
```diff
@@ -5661,6 +5662,162 @@ public void testNotHas() {
...
+            put("isOrdered", "true");
+            put("query", "multiKSQ[1]@1000");
+            put("index", "nameIdx");
+        }};
+        assertEquals(nameIdxAnnotations, nested.getAnnotations());
+
+        // satisfied by unions of two separate graph-centric queries, each satisfied by a single composite index query
+        newTx();
+        Metrics mCompMultiOr = tx.traversal().V().or(__.has("name", "bob"), __.has("weight", 100))
+            .profile().next().getMetrics(0);
```
[the remark this answers is not in the document]
Comment (farodin91) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5701]: Wouldn't be better to indent the backend-query once more?

**Comment by farodin91** [comment]
What do you think of adding a document about profiling gremlin queries and understanding profile output?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5701]: The rendering logic is controlled by DefaultTraversalMetrics in the gremlin-core package. "Or(...)" is a metrics group which consists of a few annotations ("condition=...", "orders=...", etc.) and a few subgroups ("backend-query", "optimization"). 

I think it does look a bit weird that annotations seem to have one more indent level than subgroups, but this seems to be a design choice of the gremlin package.

**Comment by li-boxuan** [comment]
> What do you think of adding a document about profiling gremlin queries and understanding profile output?

I think that's a good idea. I am planning to (if there is no other volunteer) create one or more subsequent PRs to enhance documentation on:
1) How vertex-centric index works
2) How to interpret profile output

While working on the documentation part, more fixes might be needed around profiling logic. I started this PR to fix https://github.com/JanusGraph/janusgraph/issues/2283 but ended up fixing more problems. My feeling is, more work is needed in this field.

**Comment by farodin91** [comment]
@li-boxuan Do you like to rebase? It will activate codacy.

**Comment by rngcntr** [comment]
I just discovered this PR, because @farodin91 asked me about it. Didn't have much time recently, but I think I will have the time to review it next week.

**Review by porunov — COMMENTED** [review]
Thank you @li-boxuan ! I quickly looked over this PR. The changes looks good to me in general but I just have some questions and very small nitpicks below.
I see there are some timing differences in the results you posted above. Is it just coincidence or does this PR affect performance somehow?
If it affects performance, does it affect performance or adds overhead to queries when profile step isn't present in the traversal (i.e. profiling is disabled)?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 5]
```diff
@@ -989,6 +990,70 @@ public void testConditionalIndexing() {
...
+
+        tx.addVertex("name", "bob", "prop", 100);
+        tx.commit();
+
+        // satisfied by a single graph-centric query which is satisfied by a single mixed index query
+        if (indexFeatures.supportNotQueryNormalForm()) {
+            newTx();
+            Metrics mMixedOr = tx.traversal().V().or(__.has("name", "bob"), __.has("prop", 100))
+                .profile().next().getMetrics(0);
+            assertEquals("Or(JanusGraphStep([],[name.eq(bob)]),JanusGraphStep([],[prop.eq(100)]))", mMixedOr.getName());
```
Comment (rngcntr) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1009]: That hard coded string is not very robust in case the `toString()` method for a traversal ever changes. Instead, you could initialize the traversal without the `profile()` step, save it's string representation and then apply the `profile()` step and perform the check.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 6]
```diff
@@ -1002,33 +1067,51 @@ public void testIndexSelectStrategy() {
         mixedIndex.name();
         finishSchema();
 
+        final String backend = readConfig.get(INDEX_BACKEND, INDEX);
+        String nameKey, propKey;
+        if (backend.equals("solr")) {
```
Comment (rngcntr) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1072]: Nit: Ideally, `JanusGraphIndexTest` should be entirely independent of backend implementations. Exceptions for certain backends should be implemented in the corresponding index backend test (e.g. by overriding the original test case)

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rngcntr) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5736]: `query=...` and `limit=...` belong to the sub group `backend-query`, right? Then I would prefer them to be indented the same way as annotations are indented for the top level group. I'm not sure if that is what @farodin91 actually asked for.

**Review by rngcntr — COMMENTED** [review]
I had a first look at the PR and so far, I like the improvements :+1:

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java:82]: No need to use StringBuilder for inline concatenation. It is compiled the same

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java [hunk 7]
```diff
@@ -74,8 +74,9 @@ public boolean isEmpty() {
     }
 
     @Override
-    public void observeWith(QueryProfiler profiler) {
-        queries.forEach(q -> q.observeWith(profiler));
+    public void observeWith(QueryProfiler profiler, boolean hasSiblings) {
+        boolean isSingleQuery = queries.size() == 1;
```
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java:78]: Not sure that `consistsOfMultipleQueries` is a right name because it will be `true` when it consists of 0 queries as well.
My point was that the variable is always inverted when it is used. So, I thought it will be easier if we don't invert it below. 
That said, I'm fine leaving it as is

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/KeySliceQuery.java:82]: Nice, I didn't know that

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java:78]: My initial thought was "isSingleQuery" might be easier to understand, but since doing so require multiple additional inversion operations, let me change it.
I can change it to:
`boolean consistsOfMultipleQueries = queries.size() > 1;`

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java (hunk 5 above)
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1009]: I intended to make it a change detector, to prevent unintentional changes to traverser/profiler.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java (hunk 6 above)
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1072]: Copying and pasting the whole test case does not sound elegant. Do you have advice on what helper method we could extract and override?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/testutil/JanusGraphAssert.java [hunk 8]
```diff
@@ -93,15 +94,23 @@ public static void assertIntRange(GraphTraversal<?, Integer> traversal, int star
         assertArrayEquals(intArray, traversal.toList().stream().mapToInt(i -> i).toArray());
     }
 
+    private static boolean hasBackendQuery(Metrics metrics) {
+        if (QueryProfiler.BACKEND_QUERY.equals(metrics.getName())) return true;
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/testutil/JanusGraphAssert.java:98]: Nope. It is because "backend-query" is nested under the "GraphCentricQuery" group now, i.e. one more level deeper than before.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5736]: Got it. I wondered the same but I thought that was gremlin-core's problem... let me check further.

**Comment by li-boxuan** [comment]
> I see there are some timing differences in the results you posted above. Is it just coincidence or does this PR affect performance somehow?
If it affects performance, does it affect performance or adds overhead to queries when profile step isn't present in the traversal (i.e. profiling is disabled)?

This PR doesn't affect performance. The examples I posted are very small queries whose duration has large variations.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5736]: image

From the above screenshot we can see `query=...` and `limit=...` indeed belong to the subgroup `backend-query`. I am not sure why they are not indented one more level. Maybe it's a tinkerpop bug/design.

If I add one more level (`query=...` and `limit=...` belong to the subgroup `TESTING`, which in turn belongs to the subgroup `backend-query`) it looks even more weird:
```
  GraphCentricQuery                                                                            0.000
    \_condition=(weight = 150)
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=multiKSQ[1]@1000
    \_index=weightIdx
    backend-query                                                                              0.000
      TESTING                                                          0                       0.208
    \_query=weightIdx:multiKSQ[1]@1000
    \_limit=1000
                                            >TOTAL                     -           -           0.918        -
```

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java [hunk 9]
```diff
@@ -5661,6 +5662,162 @@ public void testNotHas() {
...
+            put("isOrdered", "true");
+            put("query", "multiKSQ[1]@1000");
+            put("index", "weightIdx");
+        }};
+        assertEquals(weightIdxAnnotations, nested.getAnnotations());
+
+        // satisfied by a single graph-centric query which satisfied by intersection of two composite index queries
+        newTx();
+        Metrics mCompMultiAnd = tx.traversal().V().and(__.has("name", "bob"), __.has("weight", 100))
+            .profile().next().getMetrics(0);
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:5763]: Group "GraphCentricQuery" and its subgroup "AND-query" does not contain timing, which seems to be a bug.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/JointIndexQuery.java]: Got it, thanks. Just want to notice that `boolean isNotSingleQuery = queries.size() != 1` and `boolean consistsOfMultipleQueries = queries.size() > 1;` are different because the first one will return `true` for `0` but second will return `false` for `0`.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java (hunk 6 above)
Comment (porunov) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1080]: You don't need to copy and paste the whole method because `JanusGraphIndexTest`.
You can create a method (maybe possible `abstract` but not necessary) which returns the exact piece of information which you need for specific index backend. 
One way is to create some methods like `getNameKey()` and `getPropertyKey()` which return `"name"` and `"prop"` by default but then overwrite that method in `solr`

**Comment by li-boxuan** [comment]
As the scope of this PR grows, I feel it more difficult to debug now. To make reviewers and my life easier, I will split this PR into several smaller PRs.
